utils/s2n_fork_detection.c

```diff
@@ -0,0 +1,369 @@
+
+    if (ignore_fork_detection_for_testing == S2N_FORK_DETECT_IGNORE) {
+        /* Fork detection is meant to be disabled. Hence, return success. */
+        return S2N_RESULT_OK;
+    }
+
+    RESULT_ENSURE(fgn_state.is_fork_detection_enabled == S2N_FORK_DETECT_ENABLED, S2N_ERR_FORK_DETECTION_INIT);
+
+    /* In most cases, we would not need to increment the fork generation number.
+     * So, it is cheaper, in the expected case, to take an optimistic read lock
+     * and later aquire a write lock if needed.
+     * Note that we set the returned fgn before checking for a fork event. We
+     * need to do this because thread execution might change between releasing
+     * the read lock and taking the write lock. In that time span, another
+     * thread can reset the fork event detection sentinel and we return from
+     * s2n_get_fork_generation_number() without setting the returned fgn
+     * appropriately.
+     */
+    RESULT_ENSURE(pthread_rwlock_rdlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+    *return_fork_generation_number = fgn_state.current_fork_generation_number;
+    if (*fgn_state.zero_on_fork_addr != S2N_FORK_EVENT) {
+        /* No fork event detection. */
+        RESULT_ENSURE(pthread_rwlock_unlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+        return S2N_RESULT_OK;
+    }
+    RESULT_ENSURE(pthread_rwlock_unlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+
+    /* We are mutating the process-global, cached fork generation number. Need
+     * to acquire the write lock for that. Set returned fgn before checking the
+     * if condition with the same reasons as above.
+     */
+    RESULT_ENSURE(pthread_rwlock_wrlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+    *return_fork_generation_number = fgn_state.current_fork_generation_number;
+    if (*fgn_state.zero_on_fork_addr == S2N_FORK_EVENT) {
+        /* Fork event has been detected; reset sentinel, increment cached fork
+         * generation nunber (which is now "current" in this child process), and
+         * write incremented fork generation number to the output parameter.
+         */
+        *fgn_state.zero_on_fork_addr = S2N_NO_FORK_EVENT;
+        fgn_state.current_fork_generation_number = fgn_state.current_fork_generation_number + 1;
+        *return_fork_generation_number = fgn_state.current_fork_generation_number;
+    }
+    RESULT_ENSURE(pthread_rwlock_unlock(&fgn_state.fork_detection_rw_lock) == 0, S2N_ERR_RETRIEVE_FORK_GENERATION_NUMBER);
+
+    return S2N_RESULT_OK;
+}
+
+/* Run-time probe checking whether the system supports the MADV_WIPEONFORK fork
+ * detection mechanism.
+ *
+ * Return value:
+ *  If not supported, returns false.
+ *  If supported, returns true.
+ */
+static bool s2n_probe_madv_wipeonfork_support(void) {
+
+    bool result = false;
```

Comment:
There's a lot of logic in this method. Wouldn't you prefer to return S2N_RESULT, so you can use the safety macros and standard s2n patterns? s2n_assert_madv_wipeonfork_is_supported / s2n_is_madv_wipeonfork_supported could then translate the S2N_RESULT to a boolean for convenience.